components/sendgrid/actions/send-email/send-email.js

```diff
@@ -0,0 +1,173 @@
+      optional: true,
+    },
+    sendAt: {
+      type: "integer",
+      label: "Send At",
+      description:
+        "A unix timestamp allowing you to specify when you want your email to be delivered. This may be overridden by the `send_at` parameter set at the personalizations level. Delivery cannot be scheduled more than 72 hours in advance. If you have the flexibility, it's better to schedule mail for off-peak times. Most emails are scheduled and sent at the top of the hour or half hour. Scheduling email to avoid peak times — for example, scheduling at 10:53 — can result in lower deferral rates due to the reduced traffic during off-peak times.",
+      optional: true,
+    },
+    batchId: {
+      type: "integer",
+      label: "Batch Id",
+      description:
+        "An ID representing a batch of emails to be sent at the same time. Including a batch_id in your request allows you include this email in that batch. It also enables you to cancel or pause the delivery of that batch. For more information, see the [Cancel Scheduled Sends API](https://sendgrid.com/docs/api-reference/).",
+      optional: true,
+    },
+    asm: {
+      type: "object",
+      label: "ASM",
+      description:
+        "Advanced Suppression Manager. An object allowing you to specify how to handle unsubscribes.",
+      optional: true,
+    },
+    ipPoolName: {
+      type: "string",
+      label: "Ip Pool Name",
+      description: "The IP Pool that you would like to send this email from.",
+      optional: true,
+    },
+    mailSettings: {
+      type: "object",
+      label: "Mail Settings",
+      description:
+        "A collection of different mail settings that you can use to specify how you would like this email to be handled.",
+      optional: true,
+    },
+    trackingSettings: {
+      type: "object",
+      label: "Tracking Settings",
+      description:
+        "Settings to determine how you would like to track the metrics of how your recipients interact with your email.",
+      optional: true,
+    },
+  },
+  async run() {
+    if (
+      !this.personalizations ||
+      !this.fromEmail ||
+      !this.subject ||
+      !this.content
+    ) {
+      throw new Error(
+        "Must provide personalizations, fromEmail, subject, and content parameters."
+      );
+    }
+    const personalizations = JSON.parse(this.personalizations);
+    const content = JSON.parse(this.content);
+    const attachments = this.attachments ? JSON.parse(this.attachments) : null;
+    const categories = this.categories ? JSON.parse(this.categories) : null;
+    const replyTo = this.replyToEmail
+      ? { email: this.replyToEmail, name: this.replyToName }
+      : null;
```

Comment:
Validation is painful, so perhaps we could use [`validate.js`](http://validatejs.org/) for all these actions?

For example:
```js
const validate = require("validate.js");

module.exports = {
  run() {
    const fromEmailValidation = validate.single(this.fromEmail, {
      presence: true,
      email: true,
    });
    if (fromEmailValidation) {
      throw new Error(fromEmailValidation);
    }
  },
};
```

Reply:
Sure, I've been using `validate.js` since all these changes were requested.